gossip/gossip.go

```diff
@@ -306,11 +309,68 @@ type Callback func(key string, content []byte)
 // received. The callback method is invoked with the info key which
 // matched pattern.
 func (g *Gossip) RegisterCallback(pattern string, method Callback) {
+	if pattern == KeySystemConfig {
+		log.Warning("raw gossip callback registered on %s, consider using RegisterSystemConfigCallback",
+			KeySystemConfig)
+	}
+
 	g.mu.Lock()
 	defer g.mu.Unlock()
 	g.is.registerCallback(pattern, method)
 }
 
+// GetSystemConfig returns the local unmarshalled version of the
+// system config. It may be nil if it was never gossiped.
+func (g *Gossip) GetSystemConfig() *config.SystemConfig {
+	g.systemConfigMu.RLock()
+	defer g.systemConfigMu.RUnlock()
+	return g.systemConfig
+}
+
+type systemConfigCallback func(*config.SystemConfig)
+
+// RegisterSystemConfigCallback registers a callback for the unmarshalled
+// system config. It is called after registration, and whenever a new
+// system config is successfully unmarshalled.
+func (g *Gossip) RegisterSystemConfigCallback(method systemConfigCallback) {
+	g.systemConfigMu.Lock()
+	defer g.systemConfigMu.Unlock()
+	g.systemConfigCallbacks = append(g.systemConfigCallbacks, method)
+
+	if g.systemConfig == nil {
+		return
+	}
+
+	// Run the callback right away if we have a config.
+	// Make sure we don't use g.systemConfig after releasing the lock.
+	cfg := g.systemConfig
+	go method(cfg)
+}
+
+// updateSystemConfig is the raw gossip info callback.
+// Unmarshal the system config, and if successfuly, update out
+// copy and run the callbacks.
+func (g *Gossip) updateSystemConfig(key string, content []byte) {
+	if key != KeySystemConfig {
+		log.Fatalf("wrong key received on SystemConfig callback: %s", key)
+		return
+	}
+	cfg := &config.SystemConfig{}
+	if err := gogoproto.Unmarshal(content, cfg); err != nil {
+		log.Errorf("could not unmarshal system config on callback: %s", err)
+		return
+	}
+
+	g.systemConfigMu.Lock()
+	defer g.systemConfigMu.Unlock()
```

Comment:
move this down a line and remove the `defer`; no need to hold the lock while dispatching callbacks